### PR TITLE
feat: Implement begin device registration endpoint

### LINKED ISSUE #390: Create webAuthn create and get Http endpoint

### PR DESCRIPTION
**Pull request by talwinder50**
closes #390

Signed-off-by: talwinder.kaur <talwinder.kaur@securekey.com>

### CODE REVIEW

File: pkg/restapi/device/operations.go [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+	credential, err := o.webauthn.FinishRegistration(device, sessionData, r)
+	if err != nil {
+		common.WriteErrorResponsef(w, logger,
+			http.StatusInternalServerError, "failed to finish registration: %s", err.Error())
+
+		return
+	}
+
+	device.AddCredential(*credential)
```
Comment (llorllale) [inline pkg/restapi/device/operations.go:209]: The model relationships should be:

* 1:N between user:device
* 1:1 between device:credential

The device authenticator must authenticate with the same credential that is scoped to the website's origin (default), or the [`rpId`](https://developer.mozilla.org/en-US/docs/Web/API/PublicKeyCredentialRequestOptions/rpId).

Please remove method `AddCredential` - there should only be one credential per device.

File: pkg/restapi/device/deviceapi.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+func randomUint64() uint64 {
+	buf := make([]byte, 8)
+	rand.Read(buf)
+	return binary.LittleEndian.Uint64(buf)
+}
+
+// WebAuthnID returns the user's ID
+func (d Device) WebAuthnID() []byte {
```
Comment (llorllale) [inline pkg/restapi/device/deviceapi.go:43]: This identifier should have a clear relationship to the unique identifier currently assigned to users in our model (`User.Sub`).

Please set it to `User.Sub` for now.

File: pkg/restapi/device/operations.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+	return []common.Handler{
+		common.NewHTTPHandler(registerBeginPath, http.MethodGet, o.beginRegistration),
+		common.NewHTTPHandler(registerFinishPath, http.MethodGet, o.finishRegistration),
+	}
+}
+
+func (o *Operation) beginRegistration(w http.ResponseWriter, r *http.Request) {
+	logger.Debugf("handling device registration: %s", r.URL.String())
+
+	userSub := r.URL.Query().Get("user")
```
Comment (llorllale) [inline pkg/restapi/device/operations.go:117]: We should be fetching the user's identifier from session cookies. The session cookie should be set up at the end of `oidc.Operation.oidcCallbackHandler()`. The user's identifier will not form part of the request URL.

Opened #419 for this.

File: pkg/restapi/device/deviceapi.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+// WebAuthnCredentials returns credentials owned by the user
+func (d Device) WebAuthnCredentials() []webauthn.Credential {
+	return d.credentials
+}
+
+// CredentialExcludeList returns a CredentialDescriptor array filled
+// with all the user's credentials
+func (d Device) CredentialExcludeList() []protocol.CredentialDescriptor {
```
Comment (llorllale) [inline pkg/restapi/device/deviceapi.go:76]: This should be included in the `PublicKeyCredentialCreationOptions`.

File: pkg/restapi/device/operations.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+	"github.com/duo-labs/webauthn/webauthn"
+	"github.com/trustbloc/edge-core/pkg/log"
+	"github.com/trustbloc/edge-core/pkg/storage"
+
+	"github.com/trustbloc/edge-agent/pkg/restapi/common"
+)
+
+// Endpoints.
+const (
+	registerBeginPath  = "/register/begin/"
```
Comment (llorllale) [inline pkg/restapi/device/operations.go:29]: remove trailing backslash

File: pkg/restapi/device/operations.go (hunk 3 above)
Comment (talwinder50) [reply pkg/restapi/device/operations.go:117]: Is it okay if i address this in a separate follow up issue ?

File: pkg/restapi/device/deviceapi.go (hunk 2 above)
Comment (talwinder50) [reply pkg/restapi/device/deviceapi.go:34]: done

File: pkg/restapi/device/deviceapi.go (hunk 4 above)
Comment (talwinder50) [reply pkg/restapi/device/deviceapi.go:65]: done

File: pkg/restapi/device/operations.go (hunk 1 above)
Comment (talwinder50) [reply pkg/restapi/device/operations.go]: done

File: pkg/restapi/device/deviceapi.go (hunk 2 above)
Comment (talwinder50) [reply pkg/restapi/device/deviceapi.go:34]: done

File: pkg/restapi/device/operations.go (hunk 3 above)
Comment (talwinder50) [reply pkg/restapi/device/operations.go]: done

File: pkg/restapi/device/deviceapi.go (hunk 4 above)
Comment (talwinder50) [reply pkg/restapi/device/deviceapi.go:65]: done

File: pkg/restapi/device/operations.go (hunk 5 above)
Comment (talwinder50) [reply pkg/restapi/device/operations.go]: done

**Comment by talwinder50** [comment]
Unit test coverage is to be addressed

**Comment by talwinder50** [comment]
PR lines have increased due to go sum files. Code lines are around 1800 (sorry for it)

**Comment by talwinder50** [comment]
Patch coverage is reduced to two marshal method

**Review by rolsonquadras — APPROVED** [review]

**Review by aholovko — APPROVED** [review]

**Pull request merged**